optaplanner-core/src/main/java/org/optaplanner/core/impl/score/inliner/ScoreInliner.java

```diff
@@ -16,18 +16,122 @@
+public abstract class ScoreInliner<Score_ extends Score<Score_>, Impacter_ extends WeightedScoreImpacter> {
 
+    private final Map<String, Score_> constraintIdToWeightMap;
+    private final Map<Constraint, Impacter_> constraintToImpacterMap;
     protected final boolean constraintMatchEnabled;
+    private final Score_ zeroScore;
+    private final Map<String, DefaultConstraintMatchTotal<Score_>> constraintMatchTotalMap;
+    private final Map<Object, DefaultIndictment<Score_>> indictmentMap;
 
-    protected ScoreInliner(boolean constraintMatchEnabled) {
+    protected ScoreInliner(Map<Constraint, Score_> constraintToWeightMap, boolean constraintMatchEnabled,
+            Score_ zeroScore) {
+        this.constraintIdToWeightMap = Objects.requireNonNull(constraintToWeightMap).entrySet().stream()
+                .collect(Collectors.toMap(e -> e.getKey().getConstraintId(), Map.Entry::getValue));
+        this.constraintToImpacterMap = new HashMap<>(constraintToWeightMap.size());
         this.constraintMatchEnabled = constraintMatchEnabled;
+        this.zeroScore = zeroScore;
+        this.constraintMatchTotalMap = constraintMatchEnabled ? new LinkedHashMap<>() : null;
+        this.indictmentMap = constraintMatchEnabled ? new LinkedHashMap<>() : null;
     }
 
     public abstract Score_ extractScore(int initScore);
 
-    public abstract WeightedScoreImpacter buildWeightedScoreImpacter(Score_ constraintWeight);
+    /**
+     * Create a new instance of {@link WeightedScoreImpacter} for a particular constraint.
+     * 
+     * @param constraint never null
+     * @return never null
+     */
+    protected abstract Impacter_ buildWeightedScoreImpacter(Constraint constraint);
+
+    /**
+     * There is an impedance mismatch between {@link ScoreInliner} and CS-D.
+     *
+     * {@link WeightedScoreImpacter} is rule-specific, and therefore there needs to be an instance of it for every rule;
+     * rules are created during KieBase creation at build-time.
+     * At the same time, {@link WeightedScoreImpacter} needs to know the constraint weight,
+     * which is only available at runtime from the current working solution.
+     *
+     * These two factors together force us to create the {@link WeightedScoreImpacter} inside the rule consequence,
+     * and cache it so that each rule only creates it once.
+     *
+     * This method facilitates that. On first invocation for a particular {@link Constraint},
+     * it returns a new {@link WeightedScoreImpacter} instance.
+     * All subsequent invocations for that particular {@link Constraint}
+     * return the original {@link WeightedScoreImpacter} instance.
+     *
+     * @param constraint never null
+     * @return never null
+     */
+    public Impacter_ buildOrGetWeightedScoreImpacter(Constraint constraint) {
+        return constraintToImpacterMap.computeIfAbsent(constraint, __ -> buildWeightedScoreImpacter(constraint));
+    }
```

Comment:
Major change: ScoreInliner now keeps a cache of ScoreImpacters created from it.
This cache is necessary so that each rule can reuse its own ScoreImpacter.